### PR TITLE
Ion reader builder

### PR DESCRIPTION
**Pull request by Armanbqt**
*Issue #28 * 

Adding an IonReader Builder with IonValue as input parameter.

### CODE REVIEW

**Comment by tgregg** [comment]
Please rebase these changes onto the latest revision of `master` so that the diff doesn't include code that has already been merged.

**Comment by Armanbqt** [comment]
Done. Sorry I didn't noticed it.

File: IonDotnet/Internals/Tree/UserTreeReader.cs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        {
+            return NextHelperUser();
+        }
+        
+        public override IonType MoveNext()
+        {
+            if (!NextHelperUser())
+            {
+                _current = null;
+                return IonType.Null;
```
Comment (tgregg) [inline IonDotnet/Internals/Tree/UserTreeReader.cs]: `IonType.None`

File: IonDotnet/Internals/Tree/SystemTreeReader.cs [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+        {
+            _systemSymbols = SharedSymbolTable.GetSystem(1);
+            _current = null;
+            _eof = false;
+            _top = 0;
+            if (value.Type() == IonType.Datagram)
+            {
+                _parent = value;
+                _next = null;
+                _iter = value.GetEnumerator();
```
Comment (tgregg) [inline IonDotnet/Internals/Tree/SystemTreeReader.cs]: We should verify that this enumerator returns system values (i.e. symbol tables). In ion-java, it's possible to iterate over only the user values, or over all values including system values.

File: IonDotnet.Tests/Internals/TreeReaderTest.cs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+            var positionList = (IIonValue)_ionValueFactory.NewEmptyList();
+            positionList.Add((IIonValue)_ionValueFactory.NewInt(1234));
+            positionList.Add((IIonValue)_ionValueFactory.NewInt(5678));
+            positionList.Add((IIonValue)_ionValueFactory.NewInt(90));
+
+            var deep3 = new IonStruct {{ "deep4val", (IIonValue)_ionValueFactory.NewString("enddeep") }};
+
+            var deep2 = new IonStruct{{ "deep3", deep3 }};
+
+            var deep1 = new IonStruct { { "deep2", deep2 } };
```
Comment (tgregg) [inline IonDotnet.Tests/Internals/TreeReaderTest.cs]: The spacing is inconsistent with the previous two values ( `{ {` vs `{{` ). Also, the previous value has no space between `IonStruct` and `{`, while the others do.

File: IonDotnet/Internals/Tree/SystemTreeReader.cs [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+            _parent = (IIonValue)_stack[_top];
+            _stack[_top] = null;
+
+            // We don't know if we're at the end of the container, so check again.
+            _eof = false;
+        }
+
+        public virtual bool HasNext()
+        {
+            IonType next_type = NextHelperSystem();
```
Comment (tgregg) [inline IonDotnet/Internals/Tree/SystemTreeReader.cs]: `next_type` -> `nextType`

File: IonDotnet/Internals/Tree/UserTreeReader.cs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            if (_eof) return false;
+            if (_next != null) return true;
+
+            ClearSystemValueStack();
+
+            // read values from the system
+            // reader and if they are system values
+            // process them.  Return when we've
+            // read all the immediate system values
+            IonType next_type;
```
Comment (tgregg) [inline IonDotnet/Internals/Tree/UserTreeReader.cs]: `next_type` -> `nextType`

File: IonDotnet/Internals/Tree/SystemTreeReader.cs (hunk 2 above)
Comment (tgregg) [reply IonDotnet/Internals/Tree/SystemTreeReader.cs]: Created issue #50 to allow enumeration of system values.

**Comment by tgregg** [comment]
Closing in favor of #51

**Pull request closed without merging**